# ML Tech Interview

Hello and welcome to the Machine Learning Tech Interview. This interview will be divided in two parts: the theoretical part and the practical/coding part. 

### **I will review only the scripts that will be sent (by pull request on this repo) by 1:00 pm**

Good Luck!!

## Theoretical Part

Please answer the following questions. 

#### What are the assumptions of a linear model (or any other type of model)?

The linear model assumes there is a linear relashionship between the features and the target.
For the model to work well there must not be significant collinearity between features, no auto-correlation, multivariate normality in the distribution of residuals and the variance around the regression line is the same for the predicted values. 

#### What’s the difference between K Nearest Neighbor and K-means Clustering?

The KNN gets the nearest points in the sample (classification supervised learning), while the K-means clusters according to the mean distance of a certain group (clustering unsupervised learning). 

#### How do you address overfitting?

By getting a bigger dataset (maybe by augmenting, but it is not always the best choice); trying a different model that generalizes better, reducing complexity; do some feature regularization. 

#### Explain Naive Bayes algorithms.

It is a supervised learning classification method, that assumes independence between the variables and applies Bayes Theorem (probability of an event, based on related conditions).

#### When do you use an AUC-ROC score? What kind of information can you gather from it?

AUC measures the area underneath the ROC curve (probability curve), providing an aggregate measure of performance across all possible classification thresholds.

#### What is cross validation?

In machine learning it is used as a performance evaluation by taking different samples (splitting the data) from the dataset, training each subsample and providing all the different scores for the model.

#### What are confounding variables?

A confounding variable is an outside variable that influences the dependent and independent variable, that is not always easy to notice (lurking variable).

#### If an important metric for our company stopped appearing in our data source, how would you investigate the causes?

## Practical Machine Learning

In this challenge, you will showcase your knowledge in feature engineering, dimensionality reduction, model selection and evaluation, hyperparameter tuning, and any other techniques of machine learning.

There isn't a correct solution to this challenge. All we would like to learn is your thinking process that demonstrates your knowledge, experience, and creativity in developing machine learning models. Therefore, in addition to developing the model and optimizing its performance, you should also elaborate your thinking process and justify your decisions thoughout the iterative problem-solving process.

The suggested time to spend on this challenge is 90-120 minutes. If you don't have time to finish all the tasks you plan to do, simply document the to-dos at the end of your response.

#### Instructions:

- Download the housing prices data set (housing_prices.csv). The data is big enough to showcase your thoughts but not so that processing power will be a problem.
- Using Python, analyze the features and determine which feature set to select for modeling.
- Train and cross validate several regression models, attempting to accurately predict the SalePrice target variable.
- Evaluate all models and show comparison of performance metrics.
- State your thoughts on model performance, which model(s) you would select, and why.

#### Deliverables Checklist:

- Python code.
- Your thinking process.
- The features selected for machine learning.
- The results (e.g., performance metrics) of your selected model(s).

#### Feature analysis and selection

In [41]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.metrics import mean_squared_error
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score

pd.set_option('display.max_columns', 100)

In [2]:
# uploading data
houses = pd.read_csv('housing_prices.csv', index_col='Id')
houses.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
Id                                                                     
1          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
2          Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
3          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
4          Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
5          Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

   Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
Id                                                                       
1        Norm     1Fam     2Story            7            5       2003   
2        Norm     1Fam     1Story            6            8       1976   
3        Norm     1Fam     2Story            7            5       2001   
4        Norm     1Fam     2Story            7            5       1915   
5        Norm     1Fam     2Story            8            5       2000   

    YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                       
1           2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
2           1976     Gable  CompShg     MetalSd     MetalSd       None   
3           2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4           1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
5           2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

    MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
Id                                                                             
1        196.0        Gd        TA      PConc       Gd       TA           No   
2          0.0        TA        TA     CBlock       Gd       TA           Gd   
3        162.0        Gd        TA      PConc       Gd       TA           Mn   
4          0.0        TA        TA     BrkTil       TA       Gd           No   
5        350.0        Gd        TA      PConc       Gd       TA           Av   

   BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
Id                                                                             
1           GLQ         706          Unf           0        150          856   
2           ALQ         978          Unf           0        284         1262   
3           GLQ         486          Unf           0        434          920   
4           ALQ         216          Unf           0        540          756   
5           GLQ         655          Unf           0        490         1145   

   Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
Id                                                                             
1     GasA        Ex          Y      SBrkr       856       854             0   
2     GasA        Ex          Y      SBrkr      1262         0             0   
3     GasA        Ex          Y      SBrkr       920       866             0   
4     GasA        Gd          Y      SBrkr       961       756             0   
5     GasA        Ex          Y      SBrkr      1145      1053             0   

    GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
Id                                                                            
1        1710             1             0         2         1             3   
2        1262             0             1  

In [3]:
# NOTE: the NaNs we had before reffered to when the house didn't have a certain feature, so, the
# get_dummies dealt with it by considering it in a specific NaN column.
houses = houses.fillna(0)

In [4]:
# getting unique values length to see what object types can be coded
obj_cols = [(c, len(houses[c].unique())) for c in houses.columns if houses[c].dtype == object]
obj_cols = [c for c,v in obj_cols] # getting a list

In [5]:
# coding the objects and getting a dataset with all the features in numeric values
house_dummies = pd.get_dummies(houses[obj_cols])
houses_df = houses.drop(obj_cols, axis=1).join(house_dummies)
houses_df.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                          
1           60         65.0     8450            7            5       2003   
2           20         80.0     9600            6            8       1976   
3           60         68.0    11250            7            5       2001   
4           70         60.0     9550            7            5       1915   
5           60         84.0    14260            8            5       2000   

    YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
Id                                                                             
1           2003       196.0         706           0        150          856   
2           1976         0.0         978           0        284         1262   
3           2002       162.0         486           0        434          920   
4           1970         0.0         216           0        540          756   
5           2000       350.0         655           0        490         1145   

    1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
Id                                                                            
1        856       854             0       1710             1             0   
2       1262         0             0       1262             0             1   
3        920       866             0       1786             1             0   
4        961       756             0       1717             1             0   
5       1145      1053             0       2198             1             0   

    FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
Id                                                                             
1          2         1             3             1             8           0   
2          2         0             3             1             6           1   
3          2         1             3             1             6           1   
4          1         0             3             1             7           1   
5          2         1             4             1             9           1   

    GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
Id                                                                 
1        2003.0           2         548           0           61   
2        1976.0           2         460         298            0   
3        2001.0           2         608           0           42   
4        1998.0           3         642           0           35   
5        2000.0           3         836         192           84   

    EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  \
Id                                                                             
1               0          0            0         0        0       2    2008   
2               0          0            0         0        0       5    2007   
3               0          0            0         0        0       9    2008   
4             272          0            0         0        0       2    2006   
5               0          0            0         0        0      12    2008   

    SalePrice  MSZoning_C (all)  MSZoning_FV  MSZoning_RH  MSZoning_RL  \
Id                                                                       
1      208500                 0            0            0            1   
2      181500                 0            0            0            1   
3      223500                 0            0            0            1   
4      140000                 0            0            0            1   
5      250000                 0            0            0            1   

    MSZoning_RM  Street_Grvl  Street_Pave  Alley_0  Alley_Grvl  Alley_Pave  \
Id                                                                           
1             0            0            1        1           0           0 

In [ ]:
houses_df.info() # all numeric now and no missing values

In [37]:
houses_df.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     56.897260    57.623288   10516.828082     6.099315     5.575342   
std      42.300571    34.664304    9981.264932     1.382997     1.112799   
min      20.000000     0.000000    1300.000000     1.000000     1.000000   
25%      20.000000    42.000000    7553.500000     5.000000     5.000000   
50%      50.000000    63.000000    9478.500000     6.000000     5.000000   
75%      70.000000    79.000000   11601.500000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  \
count  1460.000000   1460.000000  1460.000000  1460.000000  1460.000000   
mean   1971.267808   1984.865753   103.117123   443.639726    46.549315   
std      30.202904     20.645407   180.731373   456.098091   161.319273   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000   
25%    1954.000000   1967.000000     0.000000     0.000000     0.000000   
50%    1973.000000   1994.000000     0.000000   383.500000     0.000000   
75%    2000.000000   2004.000000   164.250000   712.250000     0.000000   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000   

         BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  LowQualFinSF  \
count  1460.000000  1460.000000  1460.000000  1460.000000   1460.000000   
mean    567.240411  1057.429452  1162.626712   346.992466      5.844521   
std     441.866955   438.705324   386.587738   436.528436     48.623081   
min       0.000000     0.000000   334.000000     0.000000      0.000000   
25%     223.000000   795.750000   882.000000     0.000000      0.000000   
50%     477.500000   991.500000  1087.000000     0.000000      0.000000   
75%     808.000000  1298.250000  1391.250000   728.000000      0.000000   
max    2336.000000  6110.000000  4692.000000  2065.000000    572.000000   

         GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath     HalfBath  \
count  1460.000000   1460.000000   1460.000000  1460.000000  1460.000000   
mean   1515.463699      0.425342      0.057534     1.565068     0.382877   
std     525.480383      0.518911      0.238753     0.550916     0.502885   
min     334.000000      0.000000      0.000000     0.000000     0.000000   
25%    1129.500000      0.000000      0.000000     1.000000     0.000000   
50%    1464.000000      0.000000      0.000000     2.000000     0.000000   
75%    1776.750000      1.000000      0.000000     2.000000     1.000000   
max    5642.000000      3.000000      2.000000     3.000000     2.000000   

       BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  GarageYrBlt  \
count   1460.000000   1460.000000   1460.000000  1460.000000  1460.000000   
mean       2.866438      1.046575      6.517808     0.613014  1868.739726   
std        0.815778      0.220338      1.625393     0.644666   453.697295   
min        0.000000      0.000000      2.000000     0.000000     0.000000   
25%        2.000000      1.000000      5.000000     0.000000  1958.000000   
50%        3.000000      1.000000      6.000000     1.000000  1977.000000   
75%        3.000000      1.000000      7.000000     1.000000  2001.000000   
max        8.000000      3.000000     14.000000     3.000000  2010.000000   

        GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count  1460.000000  1460.000000  1460.000000  1460.000000    1460.000000   
mean      1.767123   472.980137    94.244521    46.660274      21.954110   
std       0.747315   213.804841   125.338794    66.256028      61.119149   
min       0.000000     0.000000     0.000000     0.000000       0.000000   
25%       1.000000   334.500000     0.000000     0.000000       0.000000   
50%       2.000000   480.000000     0.000000    25.000000       0.000000   
75%       2.000000   576.000000   168.000000    68.000000   

In [ ]:
# checking distribution for outliers
sns.distplot(houses_df['SalePrice'])

In [ ]:
plt.boxplot(houses_df['SalePrice'])

In [ ]:
# outliers for further use
outliers = houses_df[houses_df['SalePrice'] > 400000]

In [50]:
# getting X and y
x = houses_df.drop('SalePrice', axis=1)
y = houses_df['SalePrice']

In [8]:
# An interesting approach would be to check the features with higher correlation with the target
# for linear regression porposes
y_corr = [c for c in x.columns if abs(x[c].corr(y)) > 0.4]
len(y_corr)

26

#### Feature preparation

In [9]:
# checking collinearity
col_corr = x[y_corr].corr()[abs((x[y_corr].corr()) > 0.7) & (x[y_corr].corr() < 1)]

# getting the highly correlated features
corl = [(c, col_corr.index[col_corr[c].notnull()].tolist()) for c in col_corr.columns]
correlat = [(i,ii[0]) for i, ii in corl if len(ii)>0]
correlat = [col for pos, col in enumerate(correlat) if pos%2==0]
correlat

[('TotalBsmtSF', '1stFlrSF'),
 ('GrLivArea', 'TotRmsAbvGrd'),
 ('GarageCars', 'GarageArea')]

In [10]:
#checking which are most correlated with the target to keep, for the linear regression model
to_drop = [i if abs(x[i].corr(y)) < abs(x[ii].corr(y)) else ii for i, ii in correlat]

In [11]:
X = x[y_corr].drop(to_drop, axis=1)

In [12]:
# Another way is to do some feature reduction and scalling
features = StandardScaler().fit_transform(x[y_corr])
pca = PCA(n_components=0.99, whiten=True)

features_pca = pca.fit_transform(features)

print("Original number of features:", features.shape[1])
print("Reduced number of features:", features_pca.shape[1])
# the result is the same three I dropped, so it is good

Original number of features: 26
Reduced number of features: 23


In [13]:
# feature scaling with the dataset i will be using
scaler = StandardScaler()
X_std = scaler.fit_transform(X)

In [53]:
#train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

#### Training, Testing and Evaluating Models

In [54]:
#creating a pipeline that evaluates best polynomial size
for k in range(1,5):
    poly_model = make_pipeline (StandardScaler(), PolynomialFeatures(k), LinearRegression())
    poly_model.fit(X_train, y_train)
    y_pred = poly_model.predict(X_test)
    print(k, round(mean_squared_error(y_test, y_pred),2), poly_model.score(X_test,y_test))

1 1789010539.55 0.7798054261056999
2 3.47318161561489e+30 -4.274853216339757e+20
3 2.0748386217603533e+27 -2.553747985922101e+17
4 2.7072678129358053e+19 -3332152992.55223


In [55]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)
print('LR', round(mean_squared_error(y_test, y_pred),2), lr.score(X_test,y_test)) 

dtr = DecisionTreeRegressor()
dtr.fit(X_train, y_train)
y_pred = dtr.predict(X_test)
print('DTR', round(mean_squared_error(y_test, y_pred),2), dtr.score(X_test,y_test))  

knn = KNeighborsRegressor()
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
print('KNN', round(mean_squared_error(y_test, y_pred),2), knn.score(X_test,y_test))  

# The best model was linear regression
# checking cross validation for overfitting
print(cross_val_score(lr, X, y, cv=10), cross_val_score(dtr, X, y, cv=10), \
      cross_val_score(lr, X, y, cv=10))

LR 1789010539.55 0.7798054261056994
DTR 2740741401.75 0.6626647122693894
KNN 3039835329.66 0.6258517038754445
[0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802] [0.7023411  0.74218824 0.77440546 0.60108667 0.80237084 0.79299429
 0.73823938 0.80539489 0.45290886 0.67005473] [0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802]


In [56]:
# testing models with the full data
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1)

In [57]:
for k in range(1,8):
    poly_model = make_pipeline (PolynomialFeatures(k), LinearRegression())
    poly_model.fit(X_train, y_train)
    y_pred = poly_model.predict(X_test)
    print(k, round(mean_squared_error(y_test, y_pred),2), poly_model.score(X_test,y_test))

1 600426872.99 0.91423402631873
2 7838932987.49 -0.11972623235559543


MemoryError: 

In [58]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)
print('LR', round(mean_squared_error(y_test, y_pred),2), lr.score(X_test,y_test)) 

dtr = DecisionTreeRegressor()
dtr.fit(X_train, y_train)
y_pred = dtr.predict(X_test)
print('DTR', round(mean_squared_error(y_test, y_pred),2), dtr.score(X_test,y_test))  

knn = KNeighborsRegressor()
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
print('KNN', round(mean_squared_error(y_test, y_pred),2), knn.score(X_test,y_test)) 
# Aparently we have better results with the whole dataset
print(cross_val_score(lr, X, y, cv=10), cross_val_score(dtr, X, y, cv=10), \
      cross_val_score(lr, X, y, cv=10))

LR 600426872.99 0.9142340263184159
DTR 1212592635.63 0.8267912501045935
KNN 2415299039.3 0.6549946660330879
[0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802] [0.70951684 0.74031035 0.7969757  0.47207639 0.82048354 0.75063917
 0.75295655 0.79414927 0.68477255 0.71023481] [0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802]


In [60]:
# To check if the weird values in cross validation comes from outliers, let's use
# the dataset without them
# Dropping the outliers only from the trainning data not to influence the testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)
X_y = X_train.join(y_train)

out_drop = X_y.drop(X_y[X_y['SalePrice'] > 400000].index)

X_train = out_drop.drop('SalePrice', axis=1)
y_train = out_drop['SalePrice']

In [ ]:
for k in range(1,8):
    poly_model = make_pipeline (PolynomialFeatures(k), LinearRegression())
    poly_model.fit(X_train, y_train)
    y_pred = poly_model.predict(X_test)
    print(k, round(mean_squared_error(y_test, y_pred),2), poly_model.score(X_test,y_test))

1 1963942934.79 0.7576370955142853
2 6384001912.19 0.21217403099001875
3 2.003592129170515e+16 -2472557.643272898
4 5.7543445498287245e+17 -71012228.21523286
5 2.7307877674367524e+18 -336996376.6068025


In [61]:
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)
print('LR', round(mean_squared_error(y_test, y_pred),2), lr.score(X_test,y_test)) 

dtr = DecisionTreeRegressor()
dtr.fit(X_train, y_train)
y_pred = dtr.predict(X_test)
print('DTR', round(mean_squared_error(y_test, y_pred),2), dtr.score(X_test,y_test))  

knn = KNeighborsRegressor()
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
print('KNN', round(mean_squared_error(y_test, y_pred),2), knn.score(X_test,y_test)) 
# Aparently we have better results with the whole dataset
print(cross_val_score(lr, X, y, cv=10), cross_val_score(dtr, X, y, cv=10), \
      cross_val_score(lr, X, y, cv=10))

LR 1963942934.79 0.7576370955142845
DTR 3955400252.16 0.5118787432479215
KNN 3296725440.1 0.5931633558174513
[0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802] [0.61102521 0.66096455 0.79908542 0.55428868 0.80424896 0.80238473
 0.74243556 0.79624386 0.41743386 0.699336  ] [0.84257983 0.86727381 0.87902214 0.72720839 0.82304703 0.82697203
 0.85138253 0.83920062 0.56508307 0.87112802]


#### Conclusions

I tested several models, standardizing data, selecting different columns, taking care of collinearity with different approaches, where I found out that the best model is Linear Regression,
with all the columns. Also:
* Using PCA got worse results
* Using feature scaling got worse results
* There are some weird values in cross-validation, probably due to outliers
** I tried dropping the outliers only from the training set not to influence de evaluation of the model, but I ended up getting a worse accuracy and the same weird values at cross-validation.

<br><br>
For now i would go for the Linear Regression model using the whole dataset, but for next steps I would like to figure out what is happening at the cross validation scores where we see mostly good values near the 91% accuracy, but also there is one weird value of 0.56 which needs to be taken care of.